pkg/build/generic/docker.go

```diff
@@ -0,0 +1,100 @@
+package generic
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/testground/testground/pkg/api"
+	"github.com/testground/testground/pkg/build/common"
+	"github.com/testground/testground/pkg/docker"
+	"github.com/testground/testground/pkg/rpc"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+)
+
+const (
+	buildNetworkName = "testground-build"
+)
+
+var (
+	_ api.Builder = &DockerGenericBuilder{}
+)
+
+type DockerGenericBuilder struct {
+	Enabled bool
+}
+
+type DockerGenericBuilderConfig struct {
+	BuildArgs    map[string]*string `toml:"build_args"`
+	PushRegistry bool               `toml:"push_registry"`
+	RegistryType string             `toml:"registry_type"`
+}
+
+// Build builds a testplan written in Go and outputs a Docker container.
+func (b *DockerGenericBuilder) Build(ctx context.Context, in *api.BuildInput, ow *rpc.OutputWriter) (*api.BuildOutput, error) {
+	cfg, ok := in.BuildConfig.(*DockerGenericBuilderConfig)
+	if !ok {
+		return nil, fmt.Errorf("expected configuration type DockerGenericBuilderConfig, was: %T", in.BuildConfig)
+	}
+
+	cliopts := []client.Opt{client.FromEnv, client.WithAPIVersionNegotiation()}
+
+	var (
+		id       = in.BuildID
+		plansrc  = in.TestPlanSrcPath
+		cli, err = client.NewClientWithOpts(cliopts...)
+	)
+
+	ow = ow.With("build_id", id)
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+
+	opts := types.ImageBuildOptions{
+		Tags:        []string{id, in.BuildID},
+		BuildArgs:   cfg.BuildArgs,
+		NetworkMode: "host",
+	}
+
+	imageOpts := docker.BuildImageOpts{
+		BuildCtx:  plansrc,
```

Comment:
I'm slightly torn about this -- the base for docker:go is one level higher than this so the plan can sit next to the SDK. Maybe I should do this here as well...

The reason I decided to do it this way is that the user can run "docker build" from the plan directory and the the exact result that the daemon will build and that seems intuitive and unsurprising.

Reply:
* I think for homogeneity with other builders, we should absolutely place the plan's source under `/plan`.
* I don't think the `docker:generic` builder can support SDK linkage, because the way you link the SDK will depend entirely on the language the plan it's built with. With go, it's with `replace` directives; with JS/node/npm/Rust/Java, it will be different.

Reply:
Per this comment, I've moved the BuildCtx to basesrc rather than plansrc.

As a result, the Dockerfile has to be written slightly differently -- everything has to reference `plan/___`  rather than just plane `./` which to me seems a little unusual. Maybe this would be different if the Dockerfile were uploaded inside the plan directory. Let me know if this seems more or less confusing!